collector/bcache.go

```diff
@@ -0,0 +1,320 @@
+		allMetrics []metric
+		metrics    []metric
+	)
+
+	allMetrics = []metric{
+		// metrics in /sys/fs/bcache/<uuid>/
+		{
+			name:       "average_key_size_sectors",
+			desc:       "Average data per key in the btree (sectors).",
+			value:      float64(s.Bcache.AverageKeySize),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "btree_cache_size",
+			desc:       "Amount of memory currently used by the btree cache.",
+			value:      float64(s.Bcache.BtreeCacheSize),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "cache_available_percent",
+			desc:       "Percentage of cache device without dirty data, useable for writeback (may contain clean cached data).",
+			value:      float64(s.Bcache.CacheAvailablePercent),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "congested",
+			desc:       "Congestion.",
+			value:      float64(s.Bcache.Congested),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "root_usage_percent",
+			desc:       "Percentage of the root btree node in use (tree depth increases if too high).",
+			value:      float64(s.Bcache.RootUsagePercent),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "tree_depth",
+			desc:       "Depth of the btree.",
+			value:      float64(s.Bcache.TreeDepth),
+			metricType: prometheus.GaugeValue,
+		},
+		// metrics in /sys/fs/bcache/<uuid>/internal/
+		{
+			name:       "active_journal_entries",
+			desc:       "Number of journal entries that are newer than the index.",
+			value:      float64(s.Bcache.Internal.ActiveJournalEntries),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "btree_nodes",
+			desc:       "Depth of the btree.",
+			value:      float64(s.Bcache.Internal.BtreeNodes),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "btree_read_average_duration_nanoseconds",
+			desc:       "Average btree read duration.",
+			value:      float64(s.Bcache.Internal.BtreeReadAverageDurationNanoSeconds),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "cache_read_races",
+			desc:       "Counts instances where while data was being read from the cache, the bucket was reused and invalidated - i.e. where the pointer was stale after the read completed.",
+			value:      float64(s.Bcache.Internal.CacheReadRaces),
+			metricType: prometheus.CounterValue,
+		},
+	}
+
+	metrics = periodStatsToMetrics(&s.Bcache.Total, "total", "all")
+	allMetrics = append(allMetrics, metrics...)
+	metrics = periodStatsToMetrics(&s.Bcache.FiveMin, "5_minutes", "all")
+	allMetrics = append(allMetrics, metrics...)
+
+	for _, bdev := range s.Bdevs {
+		// metrics in /sys/fs/bcache/<uuid>/<bdev>/
+		metrics = []metric{
+			{
+				name:            "dirty_data",
+				desc:            "Amount of dirty data for this backing device in the cache.",
+				value:           float64(bdev.DirtyData),
+				metricType:      prometheus.GaugeValue,
+				extraLabel:      []string{"bdev_no"},
```

Comment:
Arguably subjective, but it's uncommon to have a `_no` suffix in such labels. In other collectors we just call it `device`, so I'd vote to use one of `device`, `bdevice`, `bdev` or `backing_device`.